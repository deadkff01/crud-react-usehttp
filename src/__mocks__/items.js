export default [
  {
    id: 1,
    name: "Nintendo Switch",
    price: 2500,
    description: "The nintendo new console",
  },
  { id: 2, name: "Playstation 4", price: 1500, description: "Sony console" },
];
