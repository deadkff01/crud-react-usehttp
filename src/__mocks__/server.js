import { rest } from "msw";
import { setupServer } from "msw/node";
import itemsMock from "__mocks__/items";

export const server = setupServer();

export const getItems = () =>
  rest.get("http://localhost:8080/get-items", (_, res, ctx) => {
    return res(ctx.status(200), ctx.json(itemsMock));
  });

export const getItemById = (id) =>
  rest.get(`http://localhost:8080/get-item-by-id/${id}`, (_, res, ctx) => {
    return res(ctx.status(200), ctx.json(itemsMock.find((i) => i.id === id)));
  });

export const getItemsError = () =>
  rest.get("http://localhost:8080/get-items", (_, res, ctx) => {
    return res(ctx.status(500), ctx.json({ message: "Internal Server Error" }));
  });
