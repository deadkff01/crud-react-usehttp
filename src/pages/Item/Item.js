import React, { useState, useEffect } from "react";
import { useParams } from "@reach/router";
import { useHttpInstance } from "services/baseUseHttp";
import Input from "components/Input";
import TextArea from "components/TextArea";
import Loader from "components/Loader";

const Item = () => {
  const params = useParams();
  const [id, setId] = useState("");
  const [name, setName] = useState("");
  const [price, setPrice] = useState("");
  const [description, setDescription] = useState("");

  const { get, put, response, loading, error } = useHttpInstance();

  useEffect(() => {
    loadItem();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  async function loadItem() {
    const item = await get(`get-item-by-id/${params.id}`);
    if (response.ok) {
      setId(item.id);
      setName(item.name);
      setPrice(item.price);
      setDescription(item.description);
    }
  }

  const updateItemRequest = async (id) => {
    const updatedItem = {
      name,
      price: Number(price),
      description,
    };
    await put(`update-item/${id}`, updatedItem);
  };

  if (error) {
    return <div>Error</div>;
  }

  if (loading) {
    return <Loader />;
  }

  return (
    <div className="mt-5 container mx-auto">
      <div className="mb-5">
        Name: <Input value={name} onChange={(e) => setName(e.target.value)} />
      </div>
      <div className="mb-5">
        Price:
        <Input
          type="number"
          value={price}
          onChange={(e) => setPrice(e.target.value)}
        />
      </div>
      <div className="mb-10">
        Description:
        <TextArea
          value={description}
          onChange={(e) => setDescription(e.target.value)}
        />
      </div>
      <button
        onClick={() => updateItemRequest(id)}
        className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded"
      >
        Update
      </button>
    </div>
  );
};

export default Item;
