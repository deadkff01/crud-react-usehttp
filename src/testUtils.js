import React from "react";
import {
  Router,
  createHistory,
  createMemorySource,
  LocationProvider,
} from "@reach/router";
import { render } from "@testing-library/react";

export function renderWithRouter(
  ui,
  { route = "/", history = createHistory(createMemorySource(route)) } = {}
) {
  return {
    ...render(<LocationProvider history={history}>{ui}</LocationProvider>),
    history,
  };
}

export function renderWithRouterParams(
  ui,
  { route = "/", history = createHistory(createMemorySource(route)) } = {}
) {
  return {
    ...render(
      <LocationProvider history={history}>
        <Router>{ui}</Router>
      </LocationProvider>
    ),
    history,
  };
}
