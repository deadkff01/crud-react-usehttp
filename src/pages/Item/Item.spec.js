import React from "react";
import { cleanup, screen } from "@testing-library/react";
import { renderWithRouterParams } from "../../testUtils";
import Item from "./Item";
import "../../setupTests";
import { server, getItemById } from "__mocks__/server";

describe("Item component", () => {
  beforeAll(() => {
    server.listen();
    jest.useRealTimers();
  });

  afterEach(() => {
    server.resetHandlers();
  });

  afterAll(() => {
    jest.useRealTimers();
    server.close();
    cleanup();
  });

  test("renders the component", async () => {
    jest.useFakeTimers();
    renderWithRouterParams(<Item path="/item/:id" />, { route: "/item/1" });

    server.use(getItemById(1));

    const element = await screen.findByTestId("container-item");
    expect(element).toMatchSnapshot();
  });
});
