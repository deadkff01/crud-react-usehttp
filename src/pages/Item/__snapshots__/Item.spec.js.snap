// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item component renders the component 1`] = `
<div
  class="mt-5 container mx-auto"
  data-testid="container-item"
>
  <div
    class="mb-5"
  >
    Name:
     
    <input
      class="css-1he0wrf"
      data-testid="input-name"
      value=""
    />
  </div>
  <div
    class="mb-5"
  >
    Price:
    <input
      class="css-1he0wrf"
      type="number"
      value=""
    />
  </div>
  <div
    class="mb-10"
  >
    Description:
    <textarea
      class="css-9jjzdi"
      rows="3"
    />
  </div>
  <button
    class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded"
  >
    Update
  </button>
</div>
`;
